main/src/ca/uwaterloo/flix/language/phase/Uncurrier2.scala

```diff
@@ -0,0 +1,366 @@
+package ca.uwaterloo.flix.language.phase
+
+import ca.uwaterloo.flix.api.Flix
+import ca.uwaterloo.flix.language.CompilationError
+import ca.uwaterloo.flix.language.ast.SimplifiedAst.{CatchRule, Def, Expression, SelectChannelRule}
+import ca.uwaterloo.flix.language.ast.ops.SimplifiedAstOps
+import ca.uwaterloo.flix.language.ast.{SimplifiedAst, Symbol, Type}
+import ca.uwaterloo.flix.util.Validation._
+import ca.uwaterloo.flix.util.{InternalCompilerException, Validation}
+
+object Uncurrier2 extends Phase[SimplifiedAst.Root, SimplifiedAst.Root] {
+
+  override def run(root: SimplifiedAst.Root)(implicit flix: Flix): Validation[SimplifiedAst.Root, CompilationError] = flix.phase("Uncurrier2") {
+    /* Uncurry an arrow type A -> (B -> C) into an arrow (A, B) -> C
```

Comment:
You have a lot closures inside `run` but I think they could just be private defs on the class itself.